### DISCUSSION TITLE
How to check if a Pile is already attached to a control before calling RentAsync/RentSync?

### DISCUSSION THREAD

**Discussion by supermomonga** [kekyo/Epoxy#46] [Q&A]
Hi, I'm using epoxy with Avalonia11.

Is there a way to check if the corresponding Pile has already been attached to the Control when calling RentAsync/RentSync?

```cs
// MainWindowViewModel.cs

public partial class MainWindowViewModel : ViewModelBase
{
    private MySettings Settings { get; set; }

    // WindowNotificationManager is Avalonia UI control.
    public Pile<WindowNotificationManager> WindowNotificationManagerPile { get; } =
        Pile.Factory.Create<WindowNotificationManager>();

    // IAsyncRelayCommand from CommunityToolkit.Mvvm
    public IAsyncRelayCommand LoadSettingsCommand { get; }

    public MainWindowViewModel()
    {
        // This will causes 'System.InvalidOperationException' which is 'You should bind an anchor before renting from the Pile.'
        LoadSettingsAsync();

        LoadSettingsCommand = new AsyncRelayCommand(async () => {
            // This will not causes error while the command is triggered by user action after view was loaded.
            await LoadSettingsAsync();
        });
    }

    private async Task LoadSettingsAsync()
    {
        t = SomePreferenceLibrary.LoadSettingsAsync();
        WindowNotificationManagerPile.RentSync(notif => notif.Show("Settings loaded"));
        MySetting = await t;
    }
}
```

Is there a way to determine if the Pile has already been attached to the control, like the following?

```cs
// Pseudo code:

// IsAttached method
if(SomePile.IsAttached) {
    SomePile.RentSync(_ => {});
}

// Try method
SomePile.TryRentSync(_ => { // Executed only when Pile already attached });

```

Or, is the practice of wanting to check whether it’s attached in the first place itself deviating from best practices?

**Comment by kekyo** [comment]
Ah, this is an interesting problem!

Surely there are situations where you want to know if the Pile is connected or not.
It is not possible to do this with the current version. However, I would like to know a bit more background.

In your example, you are explicitly trying to check if the Pile is connected, but:

* You want the Pile to be connected, or it should be connected, so you don't want to consider situations where it is not connected.
* You want to flexibly separate the process depending on whether the Pile is connected or not.

Which of the following is it, or both?

**Reply by supermomonga** [reply]
In the example I wrote, it's more like the latter.

- LoadSettingsAsync should start processing as early as possible, regardless of whether the GUI is ready.
- Upon completion of LoadSettingsAsync, if visual feedback to the user is possible, I would like to execute it (such as optionally displaying a notification toast).

However, I have started to think that a more user-friendly approach would be to queue the notification toast and ensure that it is displayed as soon as the GUI is ready...

**Reply by supermomonga** [reply]
> However, I have started to think that a more user-friendly approach would be to queue the notification toast and ensure that it is displayed as soon as the GUI is ready...

I realized that by using Queue and Fountain/Well in this way, it is not necessarily required to explicitly check the attaching state of the Pile.

```cs
// Show notification as soon as possible which queued before the GUI is ready
MainWindowWell.Add(Control.LoadedEvent, () =>
{
    WindowNotificationManagerPile.RentSync(manager =>
    {
        while(StartupNotificationQueue.TryDequeue(out var message))
        {
            manager.Show(message);
        }
    });
});
```

**Reply by kekyo** [reply]
Nice improvement! I will also consider improving availability checker interface.

**Reply by supermomonga** [reply]
Thanks!
